### PR TITLE
Add superlu_dist

### PR DESCRIPTION
**Pull request by johannesring**

### CODE REVIEW

**Comment by ahmadia** [comment]
Nice!

File: pkgs/lapack.yaml [hunk 1]
```diff
@@ -33,3 +33,7 @@ build_stages:
 - name: setup_builddir
   mode: override
   bash: |
+
+when_build_dependency:
+- {set: 'LAPACK_LDFLAGS', value: '-Wl,-rpath,${ARTIFACT}/lib -L${ARTIFACT}/lib -llapack'}
```
Comment (ahmadia) [inline pkgs/lapack.yaml]: Yes, I think this will work cross-platform, right?

File: pkgs/petsc/petsc.py [hunk 2]
```diff
@@ -108,9 +108,14 @@ def configure(ctx, stage_args):
         conf_lines.append('--with-suitesparse-include=${SUITESPARSE_DIR}/include/suitesparse')
         conf_lines.append('--with-suitesparse-lib=[${SUITESPARSE_DIR}/lib/libumfpack.a,libklu.a,libcholmod.a,libbtf.a,libccolamd.a,libcolamd.a,libcamd.a,libamd.a,libsuitesparseconfig.a]')
 
+    if 'SUPERLU_DIST' in ctx.dependency_dir_vars:
+        conf_lines.append('--with-superlu_dist=1')
+        conf_lines.append('--with-superlu_dist-include=${SUPERLU_DIST_DIR}/include/superlu_dist')
+        conf_lines.append('--with-superlu_dist-lib=[${SUPERLU_DIST_DIR}/lib/libsuperlu_dist.a]')
```
Comment (ahmadia) [inline pkgs/petsc/petsc.py]: This looks dangerous.  Won't the `.a` library be non-PIC and therefore unusable in a shared library?

**Comment by ahmadia** [comment]
This all looks good.  Let me know if you need help testing it on OS X or Linux builds, this will be very useful for the Proteus project.

File: pkgs/lapack.yaml (hunk 1 above)
Comment (johannesring) [inline pkgs/lapack.yaml]: Yes, `-Wl,rpath,` works on both Linux and OS X, while `-Wl,rpath=` does not work on OS X.

File: pkgs/petsc/petsc.py (hunk 2 above)
Comment (johannesring) [inline pkgs/petsc/petsc.py]: I'm using `-fPIC` in `make.inc.linux`, so it should be okay.

File: pkgs/superlu_dist/make.inc.linux [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+BLASDEF      = -DUSE_VENDOR_BLAS
+BLASLIB      = $(BLAS_LDFLAGS) $(LAPACK_LDFLAGS)
+METISLIB     = -Wl,-rpath,$(PARMETIS_DIR)/lib -L$(PARMETIS_DIR)/lib -lmetis
+PARMETISLIB  = -Wl,-rpath,$(PARMETIS_DIR)/lib -L$(PARMETIS_DIR)/lib -lparmetis
+LIBS         = $(DSUPERLULIB) $(BLASLIB) $(PARMETISLIB) $(METISLIB)
+ARCH         = ar
+ARCHFLAGS    = cr
+RANLIB       = ranlib
+CC           = $(MPICC)
+CFLAGS       = -O2 -fPIC
```
Comment (ahmadia) [inline pkgs/superlu_dist/make.inc.linux]: Got it, yes.  This is safe.

**Comment by johannesring** [comment]
I have only tested this on Linux until now, so it would be good if you could test it on OS X. If not, I can do it tomorrow.

**Comment by ahmadia** [comment]
I'll test it now on a Proteus build on OS X.

**Comment by ahmadia** [comment]
Looks like the tests are failing, I'll check on it later this afternoon if I have time:

```
[superlu_dist] /Users/aron/.hashdist/bld/mpi/jsrkktwepaqt/bin/mpicc  pzdrive4_ABglobal.o /Users/aron/.hashdist/tmp/superlu_dist-bkfi7r36zayx/lib/libsuperlu_dist.a -framework Accelerate  -Wl,-rpath,/Users/aron/.hashdist/bld/parmetis/b2i6ou2x7y5r/lib -L/Users/aron/.hashdist/bld/parmetis/b2i6ou2x7y5r/lib -lparmetis -Wl,-rpath,/Users/aron/.hashdist/bld/parmetis/b2i6ou2x7y5r/lib -L/Users/aron/.hashdist/bld/parmetis/b2i6ou2x7y5r/lib -lmetis -lm -o pzdrive4_ABglobal
[superlu_dist] [proxy:0:0@Arons-MacBook-Pro.local] HYDU_create_process (./utils/launch/launch.c:75): execvp error on file pddrive (No such file or directory)
[superlu_dist] [proxy:0:0@Arons-MacBook-Pro.local] HYDU_create_process (./utils/launch/launch.c:75): execvp error on file pddrive (No such file or directory)
[superlu_dist] [proxy:0:0@Arons-MacBook-Pro.local] HYDU_create_process (./utils/launch/launch.c:75): execvp error on file pddrive (No such file or directory)
[superlu_dist] [proxy:0:0@Arons-MacBook-Pro.local] HYDU_create_process (./utils/launch/launch.c:75): execvp error on file pddrive (No such file or directory)
[superlu_dist]
```

**Comment by ahmadia** [comment]
Probably the same issue we saw earlier with dynamic paths.

**Comment by ahmadia** [comment]
Nevermind, just a simple path issue.  I pushed a fix.

**Comment by ahmadia** [comment]
```
[petsc4py] /Users/aron/.hashdist/bld/mpi/jsrkktwepaqt/bin/mpicc -fPIC -Wall -Wwrite-strings -Wno-strict-aliasing -Wno-unknown-pragmas -O2 -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -DPETSC_DIR=/Users/aron/.hashdist/bld/petsc/3rx3cmpilwop -I/Users/aron/.hashdist/bld/mpi/jsrkktwepaqt/include -I/opt/X11/include -I/Users/aron/.hashdist/bld/parmetis/b2i6ou2x7y5r/include -I/Users/aron/.hashdist/bld/superlu_dist/lo6eswuegzux/include/superlu_dist -I/Users/aron/.hashdist/bld/petsc/3rx3cmpilwop/include -Isrc/include -I/Users/aron/.hashdist/bld/numpy/zqrvkl3bhcfu/lib/python2.7/site-packages/numpy/core/include -I/Users/aron/.hashdist/bld/python/ymdgbyll5gyb/include/python2.7 -c src/libpetsc4py.c -o build/temp.macosx-10.9-x86_64-2.7/src/libpetsc4py.o
[petsc4py] /Users/aron/.hashdist/bld/mpi/jsrkktwepaqt/bin/mpicc -Wl,-multiply_defined,suppress -Wl,-multiply_defined -Wl,suppress -Wl,-commons,use_dylibs -Wl,-search_paths_first -fPIC -Wall -Wwrite-strings -Wno-strict-aliasing -Wno-unknown-pragmas -O2 -bundle -undefined dynamic_lookup -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -L/Users/aron/.hashdist/bld/bzip2/gss7yln7fznj/lib -L/Users/aron/.hashdist/bld/launcher/uroqoaxazfig/lib -L/Users/aron/.hashdist/bld/ncurses/chfsm2btbjs3/lib -L/Users/aron/.hashdist/bld/openssl/sfeoc4iaaqjq/lib -L/Users/aron/.hashdist/bld/readline/bqm6dktmlhfs/lib -L/Users/aron/.hashdist/bld/sqlite/fofaqjs4h6gx/lib -L/Users/aron/.hashdist/bld/zlib/6ibn6t2utn27/lib build/temp.macosx-10.9-x86_64-2.7/src/PETSc.o build/temp.macosx-10.9-x86_64-2.7/src/libpetsc4py.o -L/Users/aron/.hashdist/bld/petsc/3rx3cmpilwop/lib -L/Users/aron/.hashdist/bld/petsc/3rx3cmpilwop/lib -lpetsc -o build/lib.macosx-10.9-x86_64-2.7/petsc4py/lib/PETSc.so
[petsc4py] ld: warning: directory not found for option '-L/Users/aron/.hashdist/bld/launcher/uroqoaxazfig/lib'
[petsc4py] ld: file not found: @rpath/libparmetis.dylib for architecture x86_64
[petsc4py] clang: error: linker command failed with exit code 1 (use -v to see invocation)
[petsc4py] error: command '/Users/aron/.hashdist/bld/mpi/jsrkktwepaqt/bin/mpicc' failed with exit status 1
[petsc4py|ERROR] Command '[u'/bin/bash', '_hashdist/build.sh']' returned non-zero exit status 1
[petsc4py|ERROR] command failed (code=1); raising
```

Still not quite working.  Sorry @johannesring - I'll need to hand this back to you.

**Comment by johannesring** [comment]
Thanks for the fix @ahmadia.

Regarding the error, I don't think it is related to this PR. I remember I had the same problem with `petsc4py` last week when I tried with `mpich` instead of `openmpi`. I tried it now on OS X and this works:

    extends:
    - file: osx.yaml
    
    packages:
      launcher:
      python:
        link: shared
      mpi:
        use: openmpi
      blas:
        use: host-osx-framework-accelerate
      lapack:
        use: host-osx-framework-accelerate
      petsc:
        build_with: |
          parmetis, scotch, suitesparse, superlu_dist
        coptflags: -O2
        link: shared
        debug: false
      petsc4py:

while it doesn't work if I switch to `mpich`. I guess the problem was introduced in #613, but I'm not sure how to fix it.

**Comment by ahmadia** [comment]
Ahh, THIS mess.

File: pkgs/superlu_dist/superlu_dist.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+dependencies:
+  build: [blas, lapack, mpi, parmetis]
+
+defaults:
+  # Skip build-time tests (not recommended)
+  without_check: false
+
+sources:
+- url: https://bitbucket.org/petsc/pkg-superlu_dist.git
+  key: git:749f33d8104157767d443ff1a1d151642751486d
```
Comment (certik) [inline pkgs/superlu_dist/superlu_dist.yaml]: This is minor, but the `key` should got first, before `url`.

**Comment by certik** [comment]
Thanks for this PR, very useful.

File: pkgs/superlu_dist/superlu_dist.yaml (hunk 4 above)
Comment (johannesring) [inline pkgs/superlu_dist/superlu_dist.yaml]: Ah, I forgot. I will fix it. Thanks.

**Comment by johannesring** [comment]
@ahmadia - the `petsc4py` error is unrelated to this PR, so I have create a different issue for this: #622.

**Comment by ahmadia** [comment]
@johannesring - Agreed.  I've made this install a little friendlier to PETSc.  If you're happy with it (I am), go ahead and merge.

**Comment by johannesring** [comment]
Thanks @ahmadia. I'm happy with it too.

**Pull request merged**